Add support for Jackson Kotlin module and improve JSON handling

kpavlov opened the pull request:
## Issue
Closes #2727

## Change

- Enhanced JSON handling by integrating `jackson-module-kotlin` for Kotlin data classes and ensuring proper module registration for serialization/deserialization.
- Refactored `JacksonJsonCodec` to enable dependency injection and added tests for better coverage and validation.
- Updated documentation and build configuration to reflect these changes.


## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] There are no breaking changes
- [x] I have added unit and/or integration tests for my change
- [ ] The tests cover both positive and negative cases
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

kpavlov committed:
Add support for Jackson Kotlin module and improve JSON handling

- Enhanced JSON handling by integrating `jackson-module-kotlin` for Kotlin data classes and ensuring proper module registration for serialization/deserialization.
- Refactored `JacksonJsonCodec` to enable dependency injection and added tests for better coverage and validation.
- Updated documentation and build configuration to reflect these changes.

kpavlov committed:
Add Javadoc comments to JacksonJsonCodec class and methods

Enhance code documentation by including detailed Javadoc comments for the JacksonJsonCodec class and its constructors. The comments explain the class purpose, usage, and configuration of the ObjectMapper for improved maintainability and clarity.

dliubarskyi commented on langchain4j-core/src/test/kotlin/dev/langchain4j/internal/JacksonJsonCodecTest.kt:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Test
+    fun `Should serialize Kotlin data classes`() {
```
this test is failing on windows:

> Expected :{\n  "status" : "ok",\n  "is_success" : true\n}
> Actual   :{\r\n  "status" : "ok",\r\n  "is_success" : true\r\n}

kpavlov replied:
Switched to [content json matchers](https://kotest.io/docs/assertions/json/content-json-matchers.html#shouldequalspecifiedjson)

dliubarskyi reviewed (commented):
@kpavlov thank you!

kpavlov committed:
Merge branch 'main' into 2727_serialize_kotlin_data_classes

kpavlov committed:
Switch to Kotest JSON-specific assertions

Replaced `shouldBe` with `shouldEqualSpecifiedJson` for improved JSON comparison in tests. Updated dependencies to include `kotest-assertions-json-jvm` and removed redundant entries for cleaner configuration.

dliubarskyi commented on langchain4j-core/pom.xml:
```diff
@@ -51,6 +51,13 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.module</groupId>
+            <artifactId>jackson-module-kotlin</artifactId>
+            <scope>runtime</scope>
```
```suggestion
            <scope>test</scope>
```

kpavlov replied:
It is runtime+optional, so will not be a required transitive dependency.
`jackson-module-kotlin` should be in the classpath in for data classes serialization to work.

dliubarskyi replied:
I understand, but in our case it is needded only for tests, so please keep it with the `test` scope.
Users will need to add it explicitly anyway, so having it runtime+optional makes no differency for them.

dliubarskyi commented on langchain4j-core/src/main/java/dev/langchain4j/internal/JacksonJsonCodec.java:
```diff
@@ -107,14 +114,36 @@ public LocalDateTime deserialize(JsonParser p, DeserializationContext ctxt) thro
                 .registerModule(module)
-                .enable(INDENT_OUTPUT);
+                .enable(INDENT_OUTPUT)
+                .enable(FAIL_ON_UNKNOWN_PROPERTIES); // To prevent issues caused by LLM hallucinations
```
isn't it enabled by default?

kpavlov replied:
Yes. I first disabled it, but later realized that it should be enabled, because unit test verifies it.
I thought, that setting it explicitly with a comment will make setup more explicit. Would you prefer to remove it?

kpavlov replied:
Removed code, keep comment

dliubarskyi replied:
No problem, feel free to keep it with explicit comment, I was just puzzled a bit

dliubarskyi commented on langchain4j/pom.xml:
```diff
@@ -42,6 +42,13 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.module</groupId>
+            <artifactId>jackson-module-kotlin</artifactId>
+            <scope>runtime</scope>
```
```suggestion
            <scope>test</scope>
```

kpavlov replied:
Same as in langchain4j-core: runtime+optional, so will not be a required transitive dependency

dliubarskyi replied:
same here

dliubarskyi committed:
Fix #2725

dliubarskyi committed:
Revert "Fix #2725"

This reverts commit b98e0a81a07610069cc9c3c458060305cbd0fb03.

dliubarskyi commented:
Sorry, commited in the wrong branch

kpavlov committed:
Merge branch 'main' into 2727_serialize_kotlin_data_classes

# Conflicts:
#	langchain4j-parent/pom.xml

kpavlov committed:
Removed explicit enabling of FAIL_ON_UNKNOWN_PROPERTIES since it is enabled by default. Adjusted test to align with updated configuration and removed redundant assertions. This ensures safer and cleaner JSON handling while avoiding unnecessary configurations.

kpavlov committed:
Update pom.xml

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

kpavlov committed:
Update pom.xml

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

kpavlov commented:
I am fine with "jackson-module-kotlin" to be test dependency

kpavlov committed:
Merge branch 'main' into 2727_serialize_kotlin_data_classes

dliubarskyi reviewed (approved):
@kpavlov thank you!
